tools/c7n_gcp/c7n_gcp/actions/notify.py

```diff
@@ -0,0 +1,153 @@
+
+from googleapiclient.errors import HttpError
+from c7n.actions import BaseNotify
+from c7n import utils
+from c7n.resolver import ValuesFrom
+from c7n_gcp.provider import resources as gcp_resources
+
+
+class Notify(BaseNotify):
+    """Example:
+        - name: gcp-notify-with-attributes
+          resource: gcp-compute
+          filters:
+           - Name: bad-instance
+          actions:
+           - type: notify
+             to:
+              - event-user
+              - resource-creator
+              - email@address
+             owner_absent_contact:
+              - other_email@address
+             # which template for the email should we use
+             template: policy-template
+             transport:
+               type: pubsub
+               topic: your-notify-topic
+    """
+    batch_size = 250
+
+    schema = {
+        'type': 'object',
+        'anyOf': [
+            {'required': ['type', 'transport', 'to']},
+            {'required': ['type', 'transport', 'to_from']}],
+        'properties': {
+            'type': {'enum': ['notify']},
+            'to': {'type': 'array', 'items': {'type': 'string'}},
+            'owner_absent_contact': {'type': 'array', 'items': {'type': 'string'}},
+            'to_from': ValuesFrom.schema,
+            'cc': {'type': 'array', 'items': {'type': 'string'}},
+            'cc_from': ValuesFrom.schema,
+            'cc_manager': {'type': 'boolean'},
+            'from': {'type': 'string'},
+            'subject': {'type': 'string'},
+            'template': {'type': 'string'},
+            'transport': {
+                'oneOf': [
+                    {'type': 'object',
+                     'required': ['type', 'topic'],
+                     'properties': {
+                         'topic': {'type': 'string'},
+                         'type': {'enum': ['pubsub']},
+                     }}],
+            },
+        }
+    }
+
+    @staticmethod
+    def register_notify_action(registry, _):
+        for resource in registry.keys():
+            klass = registry.get(resource)
+            klass.action_registry.register('notify', Notify)
+
+    def process(self, resources, event=None):
+        self.session = utils.local_session(self.manager.session_factory)
+        project = self.session.get_default_project()
+        message = {
+            'event': event,
+            'account_id': project,
+            'account': project,
+            'region': 'all',
+            'policy': self.manager.data
+        }
+
+        message['action'] = self.expand_variables(message)
+
+        for batch in utils.chunks(resources, self.batch_size):
+            message['resources'] = batch
+            receipt = self.send_data_message(message)
+            self.log.info("sent message:%s policy:%s template:%s count:%s" % (
+                receipt, self.manager.data['name'],
+                self.data.get('template', 'default'), len(batch)))
+
+    def send_data_message(self, message):
```

Comment:
why does this method exist? seems like not useful indirection, just call publish message directly.

Reply:
Removed. Again, followed pattern in aws notify. Also, originally I had a PubSubUtilities class that this method was calling to publish the message. Got rid of that and just put the publish message and other methods directly in notify based upon feedback form @thisisshi

Reply:
i wouldn't call whats in aws notify impl a pattern, it was an implementation detail specific there to supporting multiple transports. in gcp, there is only pub/sub, and if there are others it can be refactored at a later point.

Reply:
Sounds good.